### PR TITLE
Sprockets3 support

### PR DESCRIPTION
**Pull request by gafrom**
Resumbitting pull request - Tilt issue is addressed now.

Support for all Sprockets versions is enabled.
If gem is used with fresh Sprockets, it uses native Sprockets processor, otherwise it switches back to Tilt.
Tests refactored to reflect last changes.

All tests pass.

### CODE REVIEW

**Comment by aaronjensen** [comment]
Awesome, thanks. It looks like there are some conflicts (I added a `travis.yml` as well). Would you mind fixing that up?

File: sprockets-media_query_combiner.gemspec [hunk 1]
```diff
@@ -18,7 +18,7 @@ Gem::Specification.new do |gem|
 
   gem.required_ruby_version = '>= 1.9.2'
 
-  gem.add_runtime_dependency "sprockets", "~>2.1"
```
Comment (aaronjensen) [inline sprockets-media_query_combiner.gemspec]: This should also be < 4.0 probably

Comment (aaronjensen) [inline sprockets-media_query_combiner.gemspec]: or if 4.0 works then <5.0

File: lib/sprockets/media_query_combiner/version.rb [hunk 2]
```diff
@@ -1,5 +1,5 @@
 module Sprockets
   module MediaQueryCombiner
-    VERSION = "0.0.8"
```
Comment (aaronjensen) [inline lib/sprockets/media_query_combiner/version.rb]: Can you leave this out? I'll bump the version myself (typically you want to leave these out of pulls as the maintainer may want to bundle multiple things in one version... don't worry though, I'll release this as 0.0.9 as soon as its merged)

File: sprockets-media_query_combiner.gemspec (hunk 1 above)
Comment (gafrom) [inline sprockets-media_query_combiner.gemspec]: Actually, I was unable to launch Sprockets 4 on rails (it's in beta yet, and raises exceptions), but according to [this guide](https://github.com/rails/sprockets/blob/master/UPGRADING.md) I don't expect any issues running with 4.X version, thus I suggest putting `<5.0`.

File: lib/sprockets/media_query_combiner/version.rb (hunk 2 above)
Comment (gafrom) [inline lib/sprockets/media_query_combiner/version.rb]: Sure!

File: sprockets-media_query_combiner.gemspec (hunk 1 above)
Comment (aaronjensen) [inline sprockets-media_query_combiner.gemspec]: It's easy to bump it later if it does work, it's hard to pull the gem if it doesn't. I'd rather it was only whitelisted for versions it actually works against.

File: lib/sprockets/media_query_combiner/processor.rb [hunk 3]
```diff
@@ -1,15 +1,40 @@
-require 'tilt'
```
Comment (aaronjensen) [inline lib/sprockets/media_query_combiner/processor.rb]: This should probably `require "sprockets"` so that we are not relying on autoloading for `.choose`

File: lib/sprockets/media_query_combiner/processor.rb [hunk 4]
```diff
@@ -1,15 +1,40 @@
...
 module Sprockets
   module MediaQueryCombiner
-    class Processor < Tilt::Template
-      def prepare
+    
+    # serves to pick proper processor
+    class Processor
+      # returns processor class valid for current Sprockets version
+      def self.choose
+        return SprocketsProcessor if Sprockets.respond_to? :register_postprocessor
```
Comment (aaronjensen) [inline lib/sprockets/media_query_combiner/processor.rb]: Is this the only way to feature detect this? It seems totally unrelated so is no better than a version check. If there's not a better way, I'd rather just use a version check.

File: lib/sprockets/media_query_combiner/processor.rb [hunk 5]
```diff
@@ -1,15 +1,40 @@
...
+        def prepare
+        end
+
+        def evaluate(context, locals, &block)
+          Sass::MediaQueryCombiner::Combiner.combine data
+        end
+      end
+    
+    rescue LoadError
+      STDOUT.puts "Using native Sprockets processor"
```
Comment (aaronjensen) [inline lib/sprockets/media_query_combiner/processor.rb]: We should definitely be spamming stdout for things like this. Could you refactor this whole thing to conditionally define a single class based on sprockets version. There should be no `rescue` or `choose` method, you'd just use the class name. Something like:

```ruby
if # ... 
  class Processor
    # ...
  end
else
  require 'tilt'
  class Processor
    # ...
  end
end

**Comment by gafrom** [comment]
I cleared all your comments.
Also I tried to cover both Sprockets options in tests to increase coverage but to no avail.

Can you please advise on how best it can be accomplished?
I tried installing dynamicly two versions of Sprockets gem (old and fresh), but got stuck with the following error "can't activate sprockets (= 2.6.0), already activated sprockets-3.6.0" when tried to switch gems.
Thanks.

File: spec/sprockets/media_query_combiner/processor_spec.rb [hunk 6]
```diff
@@ -57,14 +54,50 @@ module Sprockets::MediaQueryCombiner
...
+      it "should work with ugly css" do
+        output[ugly_input_css].should == ugly_target_css
       end
-      combiner.evaluate(nil, nil).should ==
-        "h3{color:orange}@media (max-width:480px){h1{color:red}h2{color:blue}}@media (max-width:980px){h4{color:black}}\n"
     end
+    
   end
+
+
```
Comment (aaronjensen) [inline spec/sprockets/media_query_combiner/processor_spec.rb]: There are lots of unnecessary whitespace changes in this, the style of this gem is to have no whitespace inside classes/modules, could you remove all of it please.

**Comment by aaronjensen** [comment]
> Can you please advise on how best it can be accomplished?

I believe you'd need multiple Gemfiles to test against: https://docs.travis-ci.com/user/languages/ruby/#Testing-against-multiple-versions-of-dependencies

I'm not sure if the project is set up to do combine coverage or not. It may just work, not sure. I'm not terribly worried about it, to be honest. 

Also, now that I think about it more, given the fact that sprockets 3+ is the way of the future, I wouldn't be opposed to just supporting 3.x and bumping the major version. This gem is pretty "done" so I don't foresee needing to maintain two versions. In other words, feel free to remove the conditionals and only support 3.x if you'd rather. I'll bump a major version and release. That should keep coverage high.

**Comment by gafrom** [comment]
Yeah, dropping support for 2.x seems to be the best option. It will keep code simple and remove the need for unnecessary tests.

Thanks for time spent.

**Comment by aaronjensen** [comment]
No problem, thank you! There are still some whitespace issues which I'll mark w/ 👾 . After those are cleaned up I'll merge this in. Thanks again.

File: .travis.yml [hunk 7]
```diff
@@ -2,5 +2,7 @@ language: ruby
 before_install:
   - gem install bundler
 rvm:
+  - 1.9
   - 2.1
   - 2.2
+
```
Comment (aaronjensen) [inline .travis.yml]: 👾

File: lib/sprockets/media_query_combiner/engine.rb [hunk 8]
```diff
@@ -3,12 +3,14 @@
 if defined?(::Rails)
   module Sprockets
     module MediaQueryCombiner
+
```
Comment (aaronjensen) [inline lib/sprockets/media_query_combiner/engine.rb]: 👾

File: lib/sprockets/media_query_combiner/engine.rb [hunk 9]
```diff
@@ -3,12 +3,14 @@
...
+      
       class Engine < ::Rails::Engine
         initializer :setup_media_query_combiner, after: 'sprockets.environment', group: :all do |app|
-          app.assets.register_postprocessor 'text/css', Sprockets::MediaQueryCombiner::Processor
-          app.assets.register_bundle_processor 'text/css', Sprockets::MediaQueryCombiner::Processor
+          Sprockets.register_postprocessor    'text/css', Sprockets::MediaQueryCombiner::Processor
+          Sprockets.register_bundle_processor 'text/css', Sprockets::MediaQueryCombiner::Processor
         end
       end
+
```
Comment (aaronjensen) [inline lib/sprockets/media_query_combiner/engine.rb]: 👾

File: lib/sprockets/media_query_combiner/engine.rb [hunk 10]
```diff
@@ -3,12 +3,14 @@
 if defined?(::Rails)
   module Sprockets
     module MediaQueryCombiner
+      
       class Engine < ::Rails::Engine
         initializer :setup_media_query_combiner, after: 'sprockets.environment', group: :all do |app|
-          app.assets.register_postprocessor 'text/css', Sprockets::MediaQueryCombiner::Processor
-          app.assets.register_bundle_processor 'text/css', Sprockets::MediaQueryCombiner::Processor
+          Sprockets.register_postprocessor    'text/css', Sprockets::MediaQueryCombiner::Processor
```
Comment (aaronjensen) [inline lib/sprockets/media_query_combiner/engine.rb]: 👾  (the extra space after `register_postprocessor`)

File: spec/sprockets/media_query_combiner/processor_spec.rb (hunk 6 above)
Comment (gafrom) [inline spec/sprockets/media_query_combiner/processor_spec.rb]: Sorry, did not notice this comment

**Comment by gafrom** [comment]
Here you are, just from the oven.

Thanks again!

**Comment by aaronjensen** [comment]
Released version 1.0.0. Thanks again!

**Comment by gafrom** [comment]
Most welcome!
